motor/discovery/gcp/mql_asset_objects.go

```diff
@@ -38,9 +45,92 @@ func getTitleFamily(o gcpObject) (gcpObjectPlatformInfo, error) {
 	return gcpObjectPlatformInfo{}, errors.Newf("missing runtime info for gcp object service %s type %s", o.service, o.objectType)
 }
 
-func computeImages(m *MqlDiscovery, project string, tc *providers.Config) []*asset.Asset {
+func computeInstances(m *MqlDiscovery, project string, tc *providers.Config, sfn common.QuerySecretFn) ([]*asset.Asset, error) {
 	assets := []*asset.Asset{}
-	images := m.GetList("return gcp.project.compute.images { id name labels }")
+
+	instances, err := m.GetList("return gcp.project.compute.instances { id name labels zone { name } status networkInterfaces disks { guestOsFeatures } }")
+	if err != nil {
+		return nil, err
+	}
+	for i := range instances {
+		b := instances[i].(map[string]interface{})
+		id := b["id"].(string)
+		name := b["name"].(string)
+
+		disks := b["disks"].([]interface{})
+		if disksContainWindows(disks) {
+			log.Debug().Msgf("skipping windows instance %s", name)
+			continue
+		}
+
+		tags := b["labels"].(map[string]interface{})
+		zone := b["zone"].(map[string]interface{})
+		zoneName := zone["name"].(string)
+		status := b["status"].(string)
+		stringLabels := make(map[string]string)
+		for k, v := range tags {
+			stringLabels[k] = v.(string)
+		}
+		stringLabels["mondoo.com/instance"] = id
+
+		data, err := json.Marshal(b["networkInterfaces"])
+		if err != nil {
+			log.Error().Msgf("failed to marshal network interfaces for gcp compute instance %s", name)
+			continue
+		}
+
+		var networkIfaces []*compute.NetworkInterface
+		if err := json.Unmarshal(data, &networkIfaces); err != nil {
+			log.Error().Msgf("failed to unmarshal network interfaces for gcp compute instance %s", name)
+			continue
+		}
+
+		connections := []*providers.Config{}
+		for _, ni := range networkIfaces {
+			for _, ac := range ni.AccessConfigs {
+				if len(ac.NatIP) > 0 {
+					log.Debug().Str("instance", name).Str("ip", ac.NatIP).Msg("found public ip")
+					connections = append(connections, &providers.Config{
+						Backend:  providers.ProviderType_SSH,
+						Host:     ac.NatIP,
+						Insecure: tc.Insecure,
+					})
+				}
+			}
+		}
+
+		a := MqlObjectToAsset(project,
+			mqlObject{
+				name: name, labels: stringLabels,
+				gcpObject: gcpObject{
+					project:    project,
+					region:     zoneName,
+					name:       name,
+					id:         id,
+					service:    "compute",
+					objectType: "image",
+				},
+			}, tc)
+		a.State = mapInstanceStatus(status)
+		a.Platform.Kind = providers.Kind_KIND_VIRTUAL_MACHINE
+		a.Platform.Runtime = providers.RUNTIME_GCP_COMPUTE
+		a.Connections = connections
+		// find the secret reference for the asset
+		common.EnrichAssetWithSecrets(a, sfn)
+
+		if a.State == asset.State_STATE_RUNNING {
```

Comment:
Would it be possible to check this earlier in the code, so we can skip all the loops and data conversion when the instance isn't in a desired state?